src/libsyntax/parse/parser.rs

```diff
@@ -752,11 +759,27 @@ impl<'a> Parser<'a> {
     }
 
     pub fn parse_ident(&mut self) -> PResult<'a, ast::Ident> {
+        self.parse_ident_common(true)
+    }
+
+    fn parse_ident_common(&mut self, recover: bool) -> PResult<'a, ast::Ident> {
         match self.token {
             token::Ident(i) => {
                 if self.token.is_reserved_ident() {
-                    self.span_err(self.span, &format!("expected identifier, found {}",
-                                                      self.this_token_descr()));
+                    let mut err = self.struct_span_err(self.span,
+                                                       &format!("expected identifier, found {}",
+                                                                self.this_token_descr()));
+                    if let Some(token_descr) = self.token_descr() {
+                        err.span_label(self.span, format!("expected identifier, found {}",
+                                                          token_descr));
+                    } else {
+                        err.span_label(self.span, "expected identifier");
+                    }
```

Comment:
I don't see much point in a label that duplicates the primary message exactly.

Reply:
There have been (off hand) talks about RLS and other tools being able to reliably show only the error message _or_ the label and although the message works better in isolation, the primary span (should) works better when pointing at the code, and should be shorter as they don't have to duplicate the code's context.

I also want to accommodate people that we might have been "training" to ignore the error message in lieu of the primary label (I know _I_ have that behavior at times).

Because of those reasons, I've been trying to move away from errors that have highlighted spans devoid of labels.

Reply:
Ok, sounds reasonable.